File: termux_playwright/tunnel.py
"""Direct Socket DNS Bypass-Tunnel and In-Process CONNECT Proxy.

Bypasses Android non-root VPN / dnsproxyd blackhole by performing direct UDP
DNS queries (RFC 1035) to configurable public DNS servers (8.8.8.8, 1.1.1.1)
and proxying browser HTTPS traffic through an ephemeral in-process loopback tunnel.
"""

import asyncio
import ipaddress
import logging
import random
import socket
import struct
import threading
from typing import Optional, Dict

logger = logging.getLogger(__name__)

def is_ip_address(host: str) -> bool:
    """Check if the string is already a valid IPv4 or IPv6 address."""
    try:
        ipaddress.ip_address(host)
        return True
    except ValueError:
        return False

def query_dns_a(domain: str, dns_server: str = "8.8.8.8", port: int = 53, timeout: float = 2.0) -> Optional[str]:
    """Resolve an IPv4 address for a domain using a direct raw UDP DNS query.
    
    Adheres strictly to RFC 1035 wire format without external dependencies.
    Bypasses Bionic libc getaddrinfo and Android VPN virtual network routing.
    
    Args:
        domain: Target hostname (e.g. 'naver.com', 'uno-km.vercel.app').
        dns_server: Upstream DNS server IPv4 (default: '8.8.8.8').
        port: DNS UDP port (default: 53).
        timeout: Socket recv timeout in seconds (default: 2.0).
        
    Returns:
        IPv4 string if resolved, or None if unresolvable or timed out.
    """
    clean_domain = domain.strip().rstrip(".")
    if not clean_domain:
        return None
    if is_ip_address(clean_domain):
        return clean_domain

    try:
        q_id = random.randint(1, 65535)
        # Flags: Standard Query, Opcode 0, Recursion Desired (0x0100)
        flags = 0x0100
        header = struct.pack(">HHHHHH", q_id, flags, 1, 0, 0, 0)

        # QNAME: Sequence of labels ending with 0 byte
        qname_parts = []
        for part in clean_domain.split("."):
            encoded = part.encode("ascii")
            qname_parts.append(struct.pack("B", len(encoded)) + encoded)
        qname = b"".join(qname_parts) + b"\x00"

        # QTYPE=1 (A), QCLASS=1 (IN)
        question = qname + struct.pack(">HH", 1, 1)
        packet = header + question

        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.settimeout(timeout)
        try:
            sock.sendto(packet, (dns_server, port))
            resp, _ = sock.recvfrom(2048)
        finally:
            sock.close()

        if len(resp) < 12:
            return None

        resp_id, resp_flags, qdcount, ancount, _, _ = struct.unpack(">HHHHHH", resp[:12])
        if resp_id != q_id or ancount == 0:
            return None

        # Skip question section in response
        pos = 12 + len(question)
        for _ in range(ancount):
            if pos >= len(resp):
                break
            # Handle name field (pointer or label sequence)
            if (resp[pos] & 0xC0) == 0xC0:
                pos += 2
            else:
                while pos < len(resp) and resp[pos] != 0:
                    pos += 1 + resp[pos]
                pos += 1

            if pos + 10 > len(resp):
                break

            rtype, rclass, ttl, rdlength = struct.unpack(">HHIH", resp[pos:pos+10])
            pos += 10
            if rtype == 1 and rdlength == 4 and pos + 4 <= len(resp):
                return socket.inet_ntoa(resp[pos:pos+4])
            pos += rdlength

    except Exception as exc:
        logger.debug("Direct DNS resolution failed for '%s' via %s: %s", clean_domain, dns_server, exc)

    return None


class AsyncTunnelProxy:
    """Asynchronous loopback CONNECT proxy that intercepts browser requests
    and resolves destination domains using direct UDP DNS.
    """

    def __init__(self, host: str = "127.0.0.1", port: int = 0, dns_server: str = "8.8.8.8"):
        self.host = host
        self.port = port
        self.dns_server = dns_server
        self.server: Optional[asyncio.Server] = None
        self.assigned_port: int = 0
        self._dns_cache: Dict[str, str] = {}

    def resolve(self, domain: str) -> str:
        """Resolve domain using cache or direct DNS query."""
        if is_ip_address(domain):
            return domain
        if domain in self._dns_cache:
            return self._dns_cache[domain]
        
        resolved = query_dns_a(domain, self.dns_server)
        if resolved:
            self._dns_cache[domain] = resolved
            return resolved
        return domain

    async def _handle_client(self, client_reader: asyncio.StreamReader, client_writer: asyncio.StreamWriter):
        try:
            line = await client_reader.readline()
            if not line:
                client_writer.close()
                return

            req_line = line.decode("utf-8", errors="replace").strip()
            parts = req_line.split()
            if len(parts) >= 2 and parts[0].upper() == "CONNECT":
                host_port = parts[1]
                if ":" in host_port:
                    target_host, port_str = host_port.split(":", 1)
                    target_port = int(port_str)
                else:
                    target_host = host_port
                    target_port = 443

                # Read remaining request headers
                while True:
                    hdr = await client_reader.readline()
                    if not hdr or hdr in (b"\r\n", b"\n"):
                        break

                resolved_ip = self.resolve(target_host)
                try:
                    remote_reader, remote_writer = await asyncio.open_connection(resolved_ip, target_port)
                except Exception as conn_err:
                    logger.debug("Failed to connect to %s (%s): %s", target_host, resolved_ip, conn_err)
                    client_writer.write(b"HTTP/1.1 502 Bad Gateway\r\n\r\n")
                    await client_writer.drain()
                    client_writer.close()
                    return

                client_writer.write(b"HTTP/1.1 200 Connection Established\r\n\r\n")
                await client_writer.drain()

                async def pipe(reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
                    try:
                        while True:
                            chunk = await reader.read(32768)
                            if not chunk:
                                break
                            writer.write(chunk)
                            await writer.drain()
                    except Exception:
                        pass
                    finally:
                        try:
                            writer.close()
                        except Exception:
                            pass

                await asyncio.gather(
                    pipe(client_reader, remote_writer),
                    pipe(remote_reader, client_writer),
                    return_exceptions=True
                )
            else:
                client_writer.write(b"HTTP/1.1 405 Method Not Allowed\r\n\r\n")
                await client_writer.drain()
                client_writer.close()

        except Exception as exc:
            logger.debug("AsyncTunnelProxy connection error: %s", exc)
            try:
                client_writer.close()
            except Exception:
                pass

    async def start(self) -> int:
        """Start the async proxy server and return the listening port."""
        self.server = await asyncio.start_server(self._handle_client, self.host, self.port)
        self.assigned_port = self.server.sockets[0].getsockname()[1]
        logger.info("AsyncTunnelProxy started on %s:%d (DNS: %s)", self.host, self.assigned_port, self.dns_server)
        return self.assigned_port

    async def stop(self):
        """Gracefully terminate the async proxy server."""
        if self.server:
            self.server.close()
            await self.server.wait_closed()
            self.server = None

    def stop_sync(self):
        """Synchronously close the listening server socket."""
        if self.server:
            try:
                self.server.close()
            except Exception:
                pass
            self.server = None


class SyncTunnelProxy:
    """Synchronous wrapper for AsyncTunnelProxy that runs the event loop
    in a dedicated daemon background thread for sync Playwright launch.
    """

    def __init__(self, host: str = "127.0.0.1", port: int = 0, dns_server: str = "8.8.8.8"):
        self.host = host
        self.port = port
        self.dns_server = dns_server
        self._thread: Optional[threading.Thread] = None
        self._loop: Optional[asyncio.AbstractEventLoop] = None
        self._proxy: Optional[AsyncTunnelProxy] = None
        self.assigned_port: int = 0
        self._ready_event = threading.Event()

    def start(self) -> int:
        """Start the background proxy and return the bound port."""
        def _worker():
            self._loop = asyncio.new_event_loop()
            asyncio.set_event_loop(self._loop)
            self._proxy = AsyncTunnelProxy(self.host, self.port, self.dns_server)
            self.assigned_port = self._loop.run_until_complete(self._proxy.start())
            self._ready_event.set()
            self._loop.run_forever()

        self._thread = threading.Thread(target=_worker, daemon=True, name=f"TP-SyncTunnel-{self.dns_server}")
        self._thread.start()
        if not self._ready_event.wait(timeout=5.0):
            raise TimeoutError("SyncTunnelProxy worker thread failed to initialize within 5.0 seconds.")
        return self.assigned_port

    def stop(self):
        """Stop the background proxy thread."""
        if self._loop and self._proxy:
            try:
                if self._loop.is_running():
                    future = asyncio.run_coroutine_threadsafe(self._proxy.stop(), self._loop)
                    future.result(timeout=3.0)
            except Exception:
                pass
            try:
                if self._loop.is_running():
                    self._loop.call_soon_threadsafe(self._loop.stop)
            except Exception:
                pass
        if self._thread and self._thread.is_alive():
            self._thread.join(timeout=2.0)
        self._loop = None
        self._proxy = None

File: lib/tunnel.js
/**
 * Direct Socket DNS Bypass-Tunnel and In-Process CONNECT Proxy for Node.js
 * @license MIT
 */

'use strict';

const net = require('net');
const http = require('http');
const dgram = require('dgram');

/**
 * Resolves an IPv4 address for a domain using a direct raw UDP DNS query (RFC 1035).
 * Bypasses Android Bionic libc and VPN routing.
 * @param {string} domain Target hostname (e.g. 'uno-km.vercel.app')
 * @param {string} [dnsServer='8.8.8.8'] Upstream DNS server IPv4
 * @param {number} [port=53] DNS server UDP port
 * @param {number} [timeout=2000] Timeout in milliseconds
 * @returns {Promise<string|null>} Resolved IPv4 string or null
 */
function queryDnsA(domain, dnsServer = '8.8.8.8', port = 53, timeout = 2000) {
    return new Promise((resolve) => {
        const cleanDomain = (domain || '').trim().replace(/\.+$/, '');
        if (!cleanDomain) return resolve(null);
        if (net.isIP(cleanDomain)) return resolve(cleanDomain);

        const socket = dgram.createSocket('udp4');
        const qId = Math.floor(Math.random() * 65535) + 1;
        const flags = 0x0100; // Standard query, recursion desired

        const header = Buffer.alloc(12);
        header.writeUInt16BE(qId, 0);
        header.writeUInt16BE(flags, 2);
        header.writeUInt16BE(1, 4); // QDCOUNT
        header.writeUInt16BE(0, 6); // ANCOUNT
        header.writeUInt16BE(0, 8); // NSCOUNT
        header.writeUInt16BE(0, 10); // ARCOUNT

        const parts = cleanDomain.split('.');
        const qnameParts = [];
        for (const part of parts) {
            const lenBuf = Buffer.alloc(1);
            lenBuf.writeUInt8(part.length, 0);
            qnameParts.push(lenBuf, Buffer.from(part, 'ascii'));
        }
        qnameParts.push(Buffer.from([0]));
        const qname = Buffer.concat(qnameParts);

        const qtypeClass = Buffer.alloc(4);
        qtypeClass.writeUInt16BE(1, 0); // Type A
        qtypeClass.writeUInt16BE(1, 2); // Class IN

        const question = Buffer.concat([qname, qtypeClass]);
        const packet = Buffer.concat([header, question]);

        let timer = setTimeout(() => {
            try { socket.close(); } catch (e) {}
            resolve(null);
        }, timeout);

        socket.on('message', (msg) => {
            clearTimeout(timer);
            try {
                if (msg.length < 12) {
                    try { socket.close(); } catch (e) {}
                    return resolve(null);
                }
                const respId = msg.readUInt16BE(0);
                const ancount = msg.readUInt16BE(6);
                if (respId !== qId || ancount === 0) {
                    try { socket.close(); } catch (e) {}
                    return resolve(null);
                }

                let pos = 12 + question.length;
                for (let i = 0; i < ancount; i++) {
                    if (pos >= msg.length) break;
                    if ((msg[pos] & 0xc0) === 0xc0) {
                        pos += 2;
                    } else {
                        while (pos < msg.length && msg[pos] !== 0) {
                            pos += 1 + msg[pos];
                        }
                        pos += 1;
                    }
                    if (pos + 10 > msg.length) break;
                    const rtype = msg.readUInt16BE(pos);
                    const rdlength = msg.readUInt16BE(pos + 8);
                    pos += 10;
                    if (rtype === 1 && rdlength === 4 && pos + 4 <= msg.length) {
                        const ip = `${msg[pos]}.${msg[pos+1]}.${msg[pos+2]}.${msg[pos+3]}`;
                        try { socket.close(); } catch (e) {}
                        return resolve(ip);
                    }
                    pos += rdlength;
                }
            } catch (err) {}
            try { socket.close(); } catch (e) {}
            resolve(null);
        });

        socket.on('error', () => {
            clearTimeout(timer);
            try { socket.close(); } catch (e) {}
            resolve(null);
        });

        try {
            socket.send(packet, port, dnsServer);
        } catch (e) {
            clearTimeout(timer);
            try { socket.close(); } catch (err) {}
            resolve(null);
        }
    });
}

/**
 * In-process loopback CONNECT proxy that intercepts Chromium HTTPS requests
 * and resolves destination IPs using direct raw UDP DNS queries.
 */
class TunnelProxy {
    constructor(dnsServer = '8.8.8.8') {
        this.dnsServer = dnsServer;
        this.server = null;
        this.port = 0;
        this.cache = new Map();
        this._activeSockets = new Set();
    }

    async resolve(host) {
        if (net.isIP(host)) return host;
        if (this.cache.has(host)) return this.cache.get(host);
        const ip = await queryDnsA(host, this.dnsServer);
        if (ip) {
            this.cache.set(host, ip);
            return ip;
        }
        return host;
    }

    start() {
        return new Promise((resolve, reject) => {
            this.server = http.createServer((req, res) => {
                res.writeHead(405, { 'Content-Type': 'text/plain' });
                res.end('Method Not Allowed: TunnelProxy only accepts CONNECT requests.');
            });

            this.server.on('connect', async (req, clientSocket, head) => {
                this._activeSockets.add(clientSocket);
                clientSocket.on('close', () => this._activeSockets.delete(clientSocket));

                const [targetHost, targetPortStr] = (req.url || '').split(':');
                const targetPort = parseInt(targetPortStr || '443', 10);
                const resolvedIp = await this.resolve(targetHost);

                const remoteSocket = net.connect(targetPort, resolvedIp, () => {
                    clientSocket.write('HTTP/1.1 200 Connection Established\r\n\r\n');
                    if (head && head.length > 0) {
                        remoteSocket.write(head);
                    }
                    remoteSocket.pipe(clientSocket);
                    clientSocket.pipe(remoteSocket);
                });

                this._activeSockets.add(remoteSocket);
                remoteSocket.on('close', () => this._activeSockets.delete(remoteSocket));

                remoteSocket.on('error', () => {
                    try { clientSocket.destroy(); } catch (e) {}
                });
                clientSocket.on('error', () => {
                    try { remoteSocket.destroy(); } catch (e) {}
                });
            });

            this.server.on('error', (err) => {
                reject(err);
            });

            this.server.listen(0, '127.0.0.1', () => {
                this.port = this.server.address().port;
                resolve(this.port);
            });
        });
    }

    close() {
        for (const sock of this._activeSockets) {
            try { sock.destroy(); } catch (e) {}
        }
        this._activeSockets.clear();
        if (this.server) {
            try {
                this.server.close();
            } catch (e) {}
            this.server = null;
        }
    }
}

module.exports = {
    queryDnsA,
    TunnelProxy
};
